util/types/field_type.go

```diff
@@ -52,45 +52,58 @@ func NewFieldType(tp byte) *FieldType {
 	}
 }
 
-// String joins the information of FieldType and
-// returns a string.
-func (ft *FieldType) String() string {
-	ts := FieldTypeToStr(ft.Tp, ft.Charset)
-	ans := []string{ts}
+// CompactStr only consider Tp/CharsetBin/Flen/Deimal.
+// This is used for showing column type in infoschema.
+func (ft *FieldType) CompactStr() string {
+	ts := TypeToStr(ft.Tp, ft.Charset)
+	suffix := ""
 	switch ft.Tp {
 	case mysql.TypeEnum, mysql.TypeSet:
 		// Format is ENUM ('e1', 'e2') or SET ('e1', 'e2')
-		ans = append(ans, fmt.Sprintf("('%s')", strings.Join(ft.Elems, "','")))
+		es := make([]string, 0, len(ft.Elems))
+		for _, e := range ft.Elems {
+			e = strings.Replace(e, "'", "''", -1)
```

Comment:
mysql> create table t (c set("'e1'", "'e2'"));
Query OK, 0 rows affected (0.01 sec)

mysql> desc t;
+-------+------------------------+------+-----+---------+-------+
| Field | Type                   | Null | Key | Default | Extra |
+-------+------------------------+------+-----+---------+-------+
| c     | set('''e1''','''e2''') | YES  |     | NULL    |       |
+-------+------------------------+------+-----+---------+-------+
1 row in set (0.00 sec)